les/server_handler.go

```diff
@@ -123,47 +126,56 @@ func (h *serverHandler) handle(p *clientPeer) error {
 		p.Log().Debug("Light Ethereum handshake failed", "err", err)
 		return err
 	}
-	if p.server {
-		if err := h.server.serverset.register(p); err != nil {
-			return err
-		}
-		// connected to another server, no messages expected, just wait for disconnection
-		_, err := p.rw.ReadMsg()
-		return err
-	}
 	// Reject light clients if server is not synced.
 	if !h.synced() {
 		p.Log().Debug("Light server not synced, rejecting peer")
 		return p2p.DiscRequested
 	}
-	defer p.fcClient.Disconnect()
+	var registered bool
+	h.server.ns.Operation(func() {
+		if h.server.ns.GetField(p.Node(), clientPeerField) != nil {
+			registered = true
+		} else {
+			h.server.ns.SetFieldSub(p.Node(), clientPeerField, p)
+		}
+	})
+	if registered {
+		return errAlreadyRegistered
+	}
+
+	defer func() {
+		h.server.ns.SetField(p.Node(), clientPeerField, nil)
+		if p.fcClient != nil { // is nil when connecting another server
+			p.fcClient.Disconnect()
+		}
+	}()
+	if p.server {
+		// connected to another server, no messages expected, just wait for disconnection
```

Comment:
The `server` peer is not tracked by clientpool. If the LES-SERVER is stopped, how can we close this connection?

Reply:
Right, I totally forgot to implement this part :) Now there is a field subscription that disconnects everyone with a `clientPeerField` during `ns.Stop()` (servers have that field too). It also checks now whether the NSM is closed before accepting new peers.